### PR TITLE
EZP-28861: Allow template and controller dispatching based on matcher configuration for Content Create

### PR DESCRIPTION
**Pull request by webhdx**
> JIRA: https://jira.ez.no/browse/EZP-28861
> Requires: #215
> Requires: https://github.com/ezsystems/ezplatform-admin-ui/pull/376 (to test in AdminUI)

# TODO
- [x] rebase to master once #215 gets merged

### CODE REVIEW

File: lib/Content/View/Builder/ContentEditViewBuilder.php [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        return 'ez_content_edit:editVersionDraftAction' === $argument;
+    }
+
+    /**
+     * @param array $parameters
+     *
+     * @return \eZ\Publish\Core\MVC\Symfony\View\ContentView|\eZ\Publish\Core\MVC\Symfony\View\View
+     *
+     * @throws \eZ\Publish\Core\Base\Exceptions\InvalidArgumentException If both contentId and locationId parameters are missing
+     * @throws \eZ\Publish\Core\Base\Exceptions\UnauthorizedException
```
Comment (alongosz) [inline lib/Content/View/Builder/ContentEditViewBuilder.php]: Please type hint API exceptions when possible.
The method also throws `\eZ\Publish\API\Repository\Exceptions\NotFoundException`.

File: lib/Content/View/Builder/ContentCreateViewBuilder.php [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        return 'ez_content_edit:createWithoutDraftAction' === $argument;
+    }
+
+    /**
+     * @param array $parameters
+     *
+     * @return ContentCreateView
+     *
+     * @throws InvalidArgumentException
+     * @throws UnauthorizedException
```
Comment (alongosz) [inline lib/Content/View/Builder/ContentCreateViewBuilder.php]: \+ `@throws NotFoundException`

File: lib/Content/View/Builder/ContentCreateViewBuilder.php [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    {
+        return $this->repository->getContentLanguageService()->loadLanguage($languageCode);
+    }
+
+    /**
+     * @param $locationId
+     *
+     * @return \eZ\Publish\API\Repository\Values\Content\Location
+     */
+    private function loadLocation($locationId): Location
```
Comment (alongosz) [inline lib/Content/View/Builder/ContentCreateViewBuilder.php]: For consistency's sake:
- please hint `int` and update PhpDoc `@param`
- please add `@throws` for `NotFoundException` and `UnauthorizedException`

File: lib/Content/View/Builder/ContentCreateViewBuilder.php [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        $this->viewConfigurator->configure($view);
+
+        return $view;
+    }
+
+    /**
+     * @param string $languageCode
+     *
+     * @return Language
+     */
```
Comment (alongosz) [inline lib/Content/View/Builder/ContentCreateViewBuilder.php]: For consistency's sake please add `@throws NotFoundException`

File: lib/Content/View/Builder/ContentCreateViewBuilder.php [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        return $this->repository->getLocationService()->loadLocation($locationId);
+    }
+
+    /**
+     * @param string $contentTypeIdentifier
+     *
+     * @return ContentType
+     *
+     * @throws UnauthorizedException
+     */
```
Comment (alongosz) [inline lib/Content/View/Builder/ContentCreateViewBuilder.php]: Also `@throws NotFoundException`

File: lib/Content/View/Builder/ContentCreateViewBuilder.php [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+     *
+     * @return \eZ\Publish\API\Repository\Values\Content\Location
+     */
+    private function loadLocation($locationId): Location
+    {
+        return $this->repository->getLocationService()->loadLocation($locationId);
+    }
+
+    /**
+     * @param string $contentTypeIdentifier
```
Comment (alongosz) [inline lib/Content/View/Builder/ContentCreateViewBuilder.php]: Missing `@param array $prioritizedLanguages`

File: lib/Content/View/Builder/ContentEditViewBuilder.php [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        $this->viewParametersInjector->injectViewParameters($view, $parameters);
+        $this->viewConfigurator->configure($view);
+
+        return $view;
+    }
+
+    /**
+     * Loads Content with id $contentId.
+     *
+     * @param mixed $contentId
```
Comment (alongosz) [inline lib/Content/View/Builder/ContentEditViewBuilder.php]: Missing other params phpdoc. For `$contentId` `int` can be used.

File: lib/Content/View/Builder/ContentEditViewBuilder.php [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+     *
+     * @return \eZ\Publish\API\Repository\Values\Content\Content
+     *
+     * @throws \eZ\Publish\Core\Base\Exceptions\UnauthorizedException
```
Comment (alongosz) [inline lib/Content/View/Builder/ContentEditViewBuilder.php]: Please type hint API exception or just `UnauthorizedException`. Also throws `NotFoundException`

File: lib/Content/View/Builder/ContentEditViewBuilder.php [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+     */
+    private function loadContent($contentId, $languages = [], $versionNo = null)
```
Comment (alongosz) [inline lib/Content/View/Builder/ContentEditViewBuilder.php]: Please use PHP7 type hints for arguments and return value.

File: lib/Content/View/Builder/ContentEditViewBuilder.php [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+    {
+        return $this->repository->getContentService()->loadContent($contentId, $languages, $versionNo);
+    }
+
+    /**
+     * Loads a visible Location.
+     *
+     * @param $locationId
+     *
+     * @return \eZ\Publish\API\Repository\Values\Content\Location
```
Comment (alongosz) [inline lib/Content/View/Builder/ContentEditViewBuilder.php]: also missing throws (NotFoundException, UnauthorizedException) and type hint

File: lib/Content/View/Builder/ContentEditViewBuilder.php [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+     */
+    private function loadLocation($locationId)
```
Comment (alongosz) [inline lib/Content/View/Builder/ContentEditViewBuilder.php]: Missing PHP7 type hints for argument and return value.

File: lib/Content/View/Builder/ContentEditViewBuilder.php [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+    {
+        return $this->repository->getLocationService()->loadLocation($locationId);
```
Comment (alongosz) [inline lib/Content/View/Builder/ContentEditViewBuilder.php]: Based on other ViewBuilder classes, this makes me think it belongs to common base or trait.

File: lib/Content/View/ContentCreateView.php [hunk 13]
```diff
@@ -5,8 +5,68 @@
...
+    /**
+     * @param ContentType $contentType
+     */
+    public function setContentType(ContentType $contentType)
+    {
+        $this->contentType = $contentType;
+    }
+
+    /**
+     * @return $contentType
```
Comment (alongosz) [inline lib/Content/View/ContentCreateView.php]: return type-hint typo

File: lib/Content/View/ContentCreateView.php [hunk 14]
```diff
@@ -5,8 +5,68 @@
...
+    /**
+     * @param Location $location
+     */
+    public function setLocation(Location $location)
+    {
+        $this->location = $location;
+    }
+
+    /**
+     * @return Location
```
Comment (alongosz) [inline lib/Content/View/ContentCreateView.php]: `Location|null` or is `?Location` also correct for PhpDoc? (don't know).

File: lib/Content/View/Provider/ContentCreateView/Configured.php [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+namespace EzSystems\RepositoryForms\Content\View\Provider\ContentCreateView;
+
+use eZ\Publish\Core\MVC\Symfony\Matcher\MatcherFactoryInterface;
+use eZ\Publish\Core\MVC\Symfony\View\View;
+use eZ\Publish\Core\MVC\Symfony\View\ViewProvider;
+use EzSystems\RepositoryForms\Content\View\ContentCreateView;
+use Symfony\Component\HttpKernel\Controller\ControllerReference;
+
+/**
+ * Base for View Providers.
```
Comment (alongosz) [inline lib/Content/View/Provider/ContentCreateView/Configured.php]: Copy-paste typo.

File: lib/Content/View/Provider/ContentEditView/Configured.php [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+namespace EzSystems\RepositoryForms\Content\View\Provider\ContentEditView;
+
+use eZ\Publish\Core\MVC\Symfony\Matcher\MatcherFactoryInterface;
+use eZ\Publish\Core\MVC\Symfony\View\View;
+use eZ\Publish\Core\MVC\Symfony\View\ViewProvider;
+use EzSystems\RepositoryForms\Content\View\ContentEditView;
+use Symfony\Component\HttpKernel\Controller\ControllerReference;
+
+/**
+ * Base for View Providers.
```
Comment (alongosz) [inline lib/Content/View/Provider/ContentEditView/Configured.php]: Copy-paste typo.

**Comment by webhdx** [comment]
@mikadamczyk @Nattfarinn @ViniTou @alongosz Code review needed. Please review only 08de72e commit as it's based on #215 for the time being.

File: bundle/DependencyInjection/Compiler/ViewBuilderRegistryPass.php [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+ */
+namespace EzSystems\RepositoryFormsBundle\DependencyInjection\Compiler;
+
+use EzSystems\RepositoryForms\Content\View\Builder\ContentCreateViewBuilder;
+use EzSystems\RepositoryForms\Content\View\Builder\ContentEditViewBuilder;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+
+/**
+ * Compiler pass to register Limitation value mappers.
```
[the remark this answers is not in the document]
Comment (alongosz) [reply bundle/DependencyInjection/Compiler/ViewBuilderRegistryPass.php]: @webhdx this hasn't been addressed.

File: lib/Content/View/Builder/ContentCreateViewBuilder.php [hunk 18]
```diff
@@ -0,0 +1,273 @@
...
+    }
+
+    /**
+     * @param array $parameters
+     *
+     * @return \eZ\Publish\API\Repository\Values\Content\Location
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException
+     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException
+     * @throws \eZ\Publish\Core\Base\Exceptions\InvalidArgumentException
```
Comment (alongosz) [inline lib/Content/View/Builder/ContentCreateViewBuilder.php]: @webhdx please type-hint API.

**Pull request merged**